crucible-llvm/src/Lang/Crucible/LLVM/MemModel/MemLog.hs

```diff
@@ -75,23 +85,55 @@ import           Lang.Crucible.LLVM.MemModel.Pointer
 import           Lang.Crucible.LLVM.MemModel.Type
 import           Lang.Crucible.LLVM.MemModel.Value
 
+--------------------------------------------------------------------------------
+-- Allocations
+
 data AllocType = StackAlloc | HeapAlloc | GlobalAlloc
   deriving (Eq, Ord, Show)
 
 data Mutability = Mutable | Immutable
   deriving (Eq, Ord, Show)
 
+-- | Details of a single allocation. The @Maybe SymBV@ argument is either a
+-- size or @Nothing@ representing an unbounded allocation. The 'Mutability'
+-- indicates whether the region is read-only. The 'String' contains source
+-- location information for use in error messages.
+data AllocInfo sym =
+  forall w. (1 <= w) => AllocInfo AllocType (Maybe (SymBV sym w)) Mutability Alignment String
+
 -- | Stores writeable memory allocations.
 data MemAlloc sym
-     -- | Allocation with given block ID. The @Maybe SymBV@ argument is either a
-     -- size or @Nothing@ representing an unbounded allocation. The 'Mutability'
-     -- indicates whether the region is read-only. The 'String' contains source
-     -- location information for use in error messages.
-   = forall w. (1 <= w) => Alloc AllocType Natural (Maybe (SymBV sym w)) Mutability Alignment String
-     -- | Freeing of the given block ID.
-   | MemFree (SymNat sym) String
-     -- | The merger of two allocations.
-   | AllocMerge (Pred sym) [MemAlloc sym] [MemAlloc sym]
+    -- | A collection of consecutive basic allocations.
+  = Allocations (Map Natural (AllocInfo sym))
```

Comment:
Probably worth stating and checking the invariant (?) that these maps are always nonempty.

Reply:
good idea

Reply:
That invariant is maintained, but we don't actually rely on it except for performance reasons. So I'll document it, but I don't think it's worth checking it.

Reply:
Fixed in 77bd0b0abf766b1ce6e62d32c65004a0230aebd8.